[Handles] Design doc cleanup (#1435)
LibertyDSNP/frequency: designdocs/user-handles.md

* ```MessageSourceAccount``` exists for a user to create a handle.
* It is optional for a user to create a handle.
* Handles are unique and have a 1:1 mapping with ```MessageSourceAccount```.

> * Handles have `n`digits of suffix and is defined by a window of suffixes as `HandleSuffixMin` to `HandleSuffixMax`.

Reviewer: Space
```suggestion
* Handles have `n` digits of suffix and is defined by a window of suffixes as `HandleSuffixMin` to `HandleSuffixMax`.
```

* Ensure handles and suffixes are unique and non-conflicting.
* Make the system resistant to namespace exhaustion, homoglyphs and race conditions.
* Make the system easy to use and integrate with existing UI and wallet systems.
* Reduce the likelihood of abuse and misuse of the system for example secondary market trading of handles.

### Handles Terminology

* **Base Handle**: A base handle is a string of characters chosen by the user of Frequency.  The base handle is the part of the handle that is not suffixed.
* **Canonical Base**: A canonical base is a reduced/translated version of a base handle that is used to identify a user on the Frequency blockchain.
* **Suffix**: A suffix is a numeric value that is appended to a base handle to make it unique.  The suffix is randomly constrained and selected within a range configured on chain.
* **Delimiter**: A delimiter is a character that is used to separate the base handle and suffix.  The delimiter is a period (.) by default.
* **Display Handle**: A display handle is formed by concatenating the base handle and suffix with the delimiter.  The display handle is the handle that is displayed to the user.
  
### Handle Guidelines

> * User handles must be unique and each msa can only be mapped to one handle.

Reviewer 2: ```suggestion
* User handles must be unique and each MSA can only be mapped to one handle.
```

Reviewer 2: For consistency

* User handles must be between 3 and 20 characters (32b limit) long.

> * User handles must only contain characters that are URL safe

Reviewer: We removed this restriction
```suggestion
```

* User handles cannot contain the following substrings (or homoglyphs): @, #, :, ., ``` (this is just a suggestion, we can add more if needed)
* User handles cannot match any in a blocklist. e.g. (@admin, @everyone, @all) and others (this is just a suggestion, we can add more if needed)
* Suffixes are numeric and contain digits without any leading zeros.

> * Suffixes are u32 limited to a range defined by governance

Reviewer 3: The implementation is `u16`, no?

* Suffix will be randomly constrained.
* Homoglyph versions of handles should still resolve to the same ```msa_id``` (e.g. ```user.1234``` and ```u$er.1234``` should resolve to the same ```msa_id```).
* `Display Handle` is the handle that is displayed to the user and must be unique.

## Proposal

Handles```registry``` on Frequency chain.

### General Steps

* Create a randomly shuffled array of numeric suffixes.
* For each base handle requested, a storage map will be created with a randomly generated start index (in to the shuffled array) and a current index.  The current index will be incremented each time a new suffix is requested.
* When the current index reaches the start index, all suffixes are exhausted and an error is returned.

> * Alternativelty: Query the chain to check which values are possi available, check if the handle with the suffix is available before claiming a handle.

Reviewer: ```suggestion
* Alternativelty: Query the chain to check which values are possibly available, check if the handle with the suffix is available before claiming a handle.
```

> * Chain validates the handle and suffix and maps the handle to the ```msa_id```.

Reviewer: Making it clear the suffix is not submitted
```suggestion
* Chain validates the handle and generates the next suffix and maps the handle to the ```msa_id```.
```

> * Chain maintains a reverse mapping from handle to msa for offchain use.

Reviewer 2: ```suggestion
* Chain maintains a reverse mapping from handle to MSA for offchain use.
```

> * Chain provides a set of rpcs to query msa for a given `DisplayHandle` and vice versa.

Reviewer 2: ```suggestion
* Chain provides a set of rpcs to query MSA for a given `DisplayHandle` and vice versa.
```

* Chain providers an rpc to get presumptive suffix for a given base handle.

### Handle replay attack

To prevent replay attacks, the chain will require a mortality period for handles. This can be included in signed payload from user as a `expiration` field. The chain will check if the `expiration` is greater than the current block number. If not, the transaction will fail.  

### Handling Race Conditions

There is an exceedingly small chance that two users may attempt to claim the same handle and suffix combination within the same block. In such cases, the transaction will succeed as the distribution of suffixes by chain is random as well as on a first come basis.

> ## Sequence Diagram

Reviewer 2: Do we want to update the sequence diagram to reflect the current state of how user handles works currently? Or at least note which parts of the diagram are not implemented yet?

Author: We have all parts of sequence diag

Author: yep, updated some terms with were out dated let me know if you find anything else

* Errors -
  * `HandleAlreadyExists` if the handle already exists.
  * `InvalidHandleByteLength` if the base handle size exceeds the maximum allowed size.
  * `InvalidMessageSourceAccount` if the caller does not have a valid MSA ID.
  * `InvalidSignature` if the payload signature verification fails.
  * `InvalidSuffix` if the suffix is invalid.
  * `ProofNotValid` if the proof is not valid as of the current block.
  * `ProofHasExpired` if the proof has expired.

Validation requirements

* Handle must follow handle guidelines.
* Handle must be unique.
* Handle must be available.
* Suffix must be available.
* Suffix must be in the range of the handle.
* MSA ID must exist. Signing keys must resolve to MSA ID.
* Payload signature must be valid.
* Base handle size must not exceed the maximum allowed size.

Signature requirements

The extrinsic must be signed by the MSA owner's private key. The signature must be verified on-chain to ensure that the user is the owner of the handle. Without the possibility of a replay attack, the chain can verify that the user is the owner of the handle and no special mechanism to handle signature reuse is required.

### Retire handle

> As a network, Frequency should allow users to retire their handles. This extrinsic will allow users to retire their handles. Retired handles will be available for reuse after a time period set by governance.

Reviewer 3: This time period is currently the mortality expiration block number.

* **HandleSuffixMin**: This storage value (```u32```) will keep track of the minimum suffix value allowed for a given handle.
* **HandleSuffixMax**: This storage value (```u32```) will keep track of the maximum suffix value allowed for a given handle.
* **MSAIdToDisplayName**: This storage will map MSAIds to their corresponding user display handles.
* **CanonicalBaseHandleAndSuffixToMSAId**: This storage will map user handles to their corresponding MSAIds.
* **CanonicalBaseHandleToSuffixIndex**: This storage will map canonical version of handles to their corresponding suffix index. This is used to keep track of the current suffix index for a given canonical base.

## Required Extrinsics

### Primitives

``` rust

ClaimHandlePayload {
    base_handle: &[u8],
    expiration: u32,
}
```

### Claim handle

>  As a network, Frequency should allow users to choose their own handle, while chain will generate a random numeric suffix within the range of suffixes allowed. The display handle will be the base handle with the suffix.

Reviewer 2: ```suggestion
 As a network, Frequency should allow users to choose their own handle, while the chain will generate a random numeric suffix within the range of suffixes allowed. The display handle will be the base handle with the suffix.
```

* Base handle size must not exceed the maximum allowed size.

Signature requirements

The extrinsic must be signed by the MSA owner's private key. The signature must be verified on-chain to ensure that the user is the owner of the handle. Without the possibility of a replay attack, the chain can verify that the user is the owner of the handle and no special mechanism to handle signature reuse is required.

### Retire handle

As a network, Frequency should allow users to retire their handles. This extrinsic will allow users to retire their handles.

Note: As of this implementation

* Retiring display handle will not be available for reuse, given the unique nature of the display handle with the suffix.

Input

* origin - must be a signed origin and must be the owner of the handle.

Output

* Event - `HandleRetired` with the MSA ID and the handle.
* Errors -
  * `InvalidMessageSourceAccount` if the caller does not have a valid MSA ID.
  * `MSAHandleDoesNotExist` if the handle does not exist.
  * `HandleWithinMortalityPeriod` if an early retirement on a handle is requested

**Note:**

* Claim handle should be Capacity possible transactions.
* Retire handle should be ```Pays::No```.

> * If retiring an MSA (```retire_msa``` call), the corresponding handle should not exists.

Reviewer 2: ```suggestion
* An MSA cannot be retired (```retire_msa``` call) if a claimed user handle has not also been retired first.
```
